Actions page in HCP Terraform (#1944)
hashicorp/web-unified-docs: content/terraform-docs-common/docs/cloud-docs/actions.mdx

## Invoke an action

You can either invoke an action directly from the **Actions** page or from the invocation history page for an action. Invoking an action starts a new run. Refer to [Manage and view runs](/terraform/cloud-docs/workspaces/run/manage) for more information.

Complete the following steps to invoke an action from the **Actions** page:

1. Navigate to your workspace and click the **Actions** tab. 
1. Choose **Invoke action** from the ellipses menu.
1. When prompted, choose a run type for invoking the action. Refer to [Run modes and options](/terraform/cloud-docs/workspaces/run/modes-and-options) for more information about run types.
1. If the `action` block provisions multiple instances, you can enable the **Invoke a specific instance key** option and choose a specific action instance to invoke. 

>    Depending on your configuration, you can either choose an instance name from the **Select an instance** drop-down or specify the address of the action instance in the **Manual input** field. You can't choose an instance to invoke when the action is called from a `module` block in the root module. Refer to [Multiple action instances](#multiple-action-instances) for more information.

Author: ```suggestion
   Depending on your configuration, you can either choose an instance name from the **Select an instance** drop-down or specify the address of the action instance in the **Manual input** field. You can't invoke an action called from a `module` block when the module is configured with multiple instances. Refer to [Multiple action instances](#multiple-action-instances) for more information.
```

- **Invocation method**. A direct invocation is when a team member uses the `-invoke` flag in a [`terraform apply`](/terraform/cli/commands/apply#invoke-action) or invokes the action from the UI. A lifecycle invocation is when the [`action_trigger`](/terraform/language/meta-arguments/lifecycle#action_trigger) lifecycle argument triggers the action during a `terraform apply` operation.   
- **Source**. Choose one or more interfaces used to invoke the action.
- **Invoked by**. Only team members who have invoked the action appear in this filter.
- **Invoked**. Choose a time range. 

Either click on a run ID or choose **Go to associated run** from the ellipses menu to view the run on the [Run page](/terraform/cloud-docs/workspaces/run/manage).

Choose **View action status details** from the ellipses menu to open a tray that shows action status details. For actions invoked directly, the tray shows status for each instance of the action invoked during the run. For actions invoked as part of a resource lifecycle, the tray includes information about each resource where the action was invoked and the lifecycle rule that triggered the invocation. 

> ## Invoke an action

Reviewer: hello @trujillo-adam! 

So there's been some enhancements related to invoking actions in HCP Terraform. Mainly, the Invoke Action modal now allows users to choose a specific action instance if the action was declared with `count` or `for_each` meta-argument (see image below). 

<img width="622" height="508" alt="Image" src="https://github.com/user-attachments/assets/3274c530-cfb3-4402-95b1-d5cb5393b982" />

With that added support, there are a few related limitations around which action instances the modal can automatically detect and present to users. I'm not sure how we typically add or explain these types of limitations in the docs, so I wanted to share the details here so we can decide where they belong and how much depth would be useful.

here are the limitation details:

**Module Action limitation:**
- If an action is declared inside a `module` block, and that `module` block uses the `count` or `for_each` meta-argument, users will not be able to select a specific module instance from the Invoke Action modal.
  - For example, the modal cannot offer module instance options such as `module.foo["east"]` or `module.foo[0]`. 
  - To get around this limitation, users can invoke the action through the CLI or API with the full action address, including the module instance key, or use **Re-invoke action** from a previous direct invocation when available.
  - We’re also rendering an alert in the modal to warn users that module blocks with `count` or `for_each` may not be invocable from the UI. (see image below) 

<img width="611" height="473" alt="Image" src="https://github.com/user-attachments/assets/0427a7a3-2b60-47c7-b8f0-8bce53fbe2a2" />

<br/><br/>

**Invoke Action modal limitations:**
- While the Invoke Action modal can show action instance options for simple `count` or `for_each` meta-arguments (ex. literal values like `count = 3` or a single direct variable reference like `for_each = var.foo`), more complex expressions **require users to manually** enter the action instance address. 
  - Examples include `local.*`, nested variable traversals like `var.obj.attr`, multiple-reference expressions like `var.a + var.b` or `merge(var.map_a, var.map_b)`, and references such as `each.*`, `self.*`, resource references, or module output references.
- Function-wrapped literal collections are not recognized by the Invoke Action modal. For example, `for_each = toset(["east", "west"])` or `for_each = tomap({k1="v1", k2="v2"})` does not show an instance dropdown or a manual instance input field. 
  - Note: For function-wrapped direct variable references, such as `toset(var.regions)`, the modal will still present the user with the list of available instance options when the values can be resolved.

I’m not sure if these belong in the docs too, but there are a couple of other limitations that can be avoided with good authoring practices:
- Actions nested more than 20 module calls deep are not available for invocation from the Invoke Action modal.
- The Invoke Action modal does not show dropdown instance options when a resolved `for_each` action has more than 100 keys. Users can still manually enter the full action instance address, invoke through CLI/API, or use **Re-invoke action** from a previous direct invocation when available.

---

> page_description: Manage Terraform actions

Reviewer 2: ```suggestion
page_title: Manage Terraform actions
```

last_modified: 2026-03-06T00:41:01.000Z
# END AUTO GENERATED METADATA
---

# Manage Terraform actions

This topic describes how to find, view, invoke, and track Terraform actions, including their invocation history, in HCP Terraform. Use the **Actions** page to help you find potential issues related to adhoc operations performed by an invoked action.

@include 'tfc-package-callouts/standard.mdx'

## Introduction

Actions are preset operations built into providers that you can invoke with Terraform. Actions can trigger automations outside of Terraform, such as running Ansible playbooks and Lambda jobs. Actions do not affect resource state. Refer to [Invoke an action](/terraform/language/invoke-actions) in the configuration language documentation for more information.

HCP Terraform includes a dedicated **Actions** page for viewing and invoking actions declared in your workspace configuration. The **Actions** page helps users gain insight about any ad hoc operations that have been performed.

## Requirements

- Terraform version 1.14 or newer. Refer to the [Terraform version settings](/terraform/cloud-docs/workspaces/settings#terraform-version) for more information. 

> - You must be on an IBM HCP Terraform standard plan to access the **Actions** page. Refer to the [pricing page](https://www.hashicorp.com/en/pricing?tab=terraform) for more information.

Reviewer 3: Hm. So here, we call it an "**IBM** HCP Terraform ..." plan. We might want to make this reference uniform throughout, then.

- **Invocation method**. A direct invocation is when a team member uses the `-invoke` flag in a [`terraform apply`](/terraform/cli/commands/apply#invoke-action) or invokes the action from the UI. A lifecycle invocation is when the [`action_trigger`](/terraform/language/meta-arguments/lifecycle#action_trigger) lifecycle argument triggers the action during a `terraform apply` operation.   
- **Source**. Choose one or more interfaces used to invoke the action.
- **Invoked by**. Only team members who have invoked the action appear in this filter.
- **Invoked**. Choose a time range. 

Click on an action to [view its invocation history](#view-invocation-history). You can also open the ellipses menu to [invoke an action](#invoke-an-action).

## View invocation history

> Navigate to your workspace and click the **Actions** in the sidebar. Click on an action in the **Name** column to view its details and invocation history by run ID. 

Reviewer 3: Nit - I would remove the first '`the`' in "`...click the **Actions** in the sidebar.`"

Author: ```suggestion
Navigate to your workspace and click **Actions** in the sidebar. Click on an action in the **Name** column to view its details and invocation history by run ID. 
```

The **Summary** tile presents the following information from the registry:

- description
- provider
- a link to the documentation

The **Summary** tile also shows how many times the action has been invoked in the workspace. 

### Search and filter action invocations

You can sort the invocation history by last invocation. Use the search bar to find a specific run ID or choose filters from the **Filters** menu to narrow down search results. You can apply the following types of filters:

- **Action status**. Choose one or more statuses. Because an action may execute successfully even if the run fails, the action status may not match the state of the associated run. HCP Terraform tracks the action status for each invocation, which may execute multiple times per run or across runs. As a result, HCP Terraform reports the status as **Run in progress** when one instance is still in progress or **Not successful** when one instance ran unsuccessfully. 

> - **Invocation method**. A direct invocation is when a team member uses the `-invoke` flag in a [`terraform apply`](/terraform/cli/commands/apply#invoke-action). A lifecycle invocation is when the [`action_trigger`](/terraform/language/meta-arguments/lifecycle#action_trigger) lifecycle argument triggers the action during a `terraform apply` operation.   

Author: ```suggestion
- **Invocation method**. A direct invocation is when a team member uses the `-invoke` flag in a [`terraform apply`](/terraform/cli/commands/apply#invoke-action) or invokes the action from the UI. A lifecycle invocation is when the [`action_trigger`](/terraform/language/meta-arguments/lifecycle#action_trigger) lifecycle argument triggers the action during a `terraform apply` operation.   
```

- **Invoked by**. Only team members who have invoked the action appear in this filter.
- **Invoked**. Choose a time range. 

Either click on a run ID or choose **Go to associated run** from the ellipses menu to view the run on the [Run page](/terraform/cloud-docs/workspaces/run/manage).

Choose **View action status details** from the ellipses menu to open a tray that shows action status details. For actions invoked directly, the tray shows status for each instance of the action invoked during the run. For actions invoked as part of a resource lifecycle, the tray includes information about each resource where the action was invoked and the lifecycle rule that triggered the invocation. 

## Invoke an action

You can either invoke an action directly from the **Actions** page or from the invocation history page for an action. Invoking an action starts a new run. Refer to [Manage and view runs](/terraform/cloud-docs/workspaces/run/manage) for more information.

Complete the following steps to invoke an action from the **Actions** page:

1. Navigate to your workspace and click the **Actions** tab. 
1. Choose **Invoke action** from the ellipses menu.

> 1. When prompted, choose a run type for invoking the action. Refer to [Run modes and options(/terraform/cloud-docs/workspaces/run/modes-and-options)] for more information about run types.

Reviewer 3: Looks like this link has the closing `]` bracket in the wrong spot

Author: ```suggestion
1. When prompted, choose a run type for invoking the action. Refer to [Run modes and options](/terraform/cloud-docs/workspaces/run/modes-and-options) for more information about run types.
```
